Fix error `Model not defined`

devmotion opened the pull request:
Fixes the error
```julia
ERROR: LoadError: LoadError: LoadError: UndefVarError: Model not defined
Stacktrace:
 [1] top-level scope at /home/david/.julia/packages/DynamicPPL/e9MX7/src/compiler.jl:449
 [2] include(::String) at ./client.jl:439
 [3] top-level scope at /home/david/.julia/dev/Turing/test/test_utils/AllUtils.jl:7
 [4] include(::String) at ./client.jl:439
 [5] top-level scope at /home/david/.julia/dev/Turing/test/runtests.jl:9
 [6] include(::String) at ./client.jl:439
 [7] top-level scope at none:6
in expression starting at /home/david/.julia/dev/Turing/test/test_utils/models.jl:2
in expression starting at /home/david/.julia/dev/Turing/test/test_utils/AllUtils.jl:7
in expression starting at /home/david/.julia/dev/Turing/test/runtests.jl:9
ERROR: Package Turing errored during testing
```
when trying to use Turing with the master branch of DynamicPPL. I'm not sure if there's a good and reliable way to catch this error in the CI tests - I guess Model is always defined since we load DynamicPPL explicitly, isn't it?

devmotion commented:
BTW actually I think we should always use `$DynamicPPL.Model` etc. instead of `DynamicPPL.Model` in macro expressions. Otherwise it errors if only `DynamicPPL.Model` is not imported (can happen if, e.g., one specifies only `using DynamicPPL: VarInfo`).

A simple example illustrates this:
```julia
julia> module A
       module B
       f() = rand()
       macro g()
           esc(:(B.f()))
       end
       macro h()
           esc(:($B.f()))
       end
       end
       end
Main.A

julia> A.B.f()
0.37300768088145975

julia> A.B.@g()
ERROR: UndefVarError: B not defined
Stacktrace:
 [1] top-level scope at REPL[3]:1

julia> A.B.@h()
0.7635233525451748
```

devmotion commented:
Oh actually I think one should define
```julia
macro h()
    esc(:($f())
end
```
or
```julia
macro h()
    esc(:($(B.f)())
end
```
Ideally, however in this would be
```julia
macro h()
    :(f())
end
```
and only escaping user input.

mohamed82008 commented:
I think if we don't call `using DynamicPPL` in the tests (by running Turing tests in a separate module perhaps) will give us "the full Turing experience" when testing.

devmotion commented:
When building the function body and the arguments, we have to use `esc` and `gensym`, but in the other cases it could be removed. I replaced the `@isassumption` macro with a function and removed the superfluous branches if the LHS is not an argument of the model - even though they should be compiled away, including them generates incorrect source code since variables are referenced before they are initialized.

mohamed82008 commented:
I will give this another look tomorrow.

mohamed82008 commented:
We also need to update the compiler docs in the Turing website to reflect the new changes in DPPL.

phipsgabler commented on src/compiler.jl:
```diff
@@ -294,45 +293,58 @@ function generate_tilde(left, right, model_info)
+            @gensym isassumption
+            return quote
+                $expr
+                $isassumption = $(DynamicPPL.isassumption(model, left))
```
Hah.  I like this.  Macros have spoilt me.

phipsgabler commented on src/compiler.jl:
```diff
@@ -347,46 +359,58 @@ function generate_dot_tilde(left, right, model_info)
-            )
+
+        return quote
+            $expr
```
If `expr` were made a `Vector{Expr}`, then you could write `$(expr...)` here and an in the other results, and get rid of the nested block in the generated code.  Not changing any semantics, but I'm always for more easily readable macro outputs.

devmotion replied:
Good idea, I'll change it.

devmotion commented on src/compiler.jl:
```diff
@@ -294,45 +293,54 @@ function generate_tilde(left, right, model_info)
+
+    @gensym tmpright
+    top = [:($tmpright = $right),
+           :($(DynamicPPL.assert_dist)($tmpright, msg = $(wrong_dist_errormsg(@__LINE__))))]
```
BTW the use of `@__LINE__` seems a bit useless to me. It will just point to the line number in compiler.jl (i.e., it just prints "... on line 299") but not to any line numbers in the user code. I don't think this is helpful for any user?

devmotion replied:
Maybe it's more useful to use `__source__.line` (if possible)?

devmotion replied:
https://docs.julialang.org/en/v1/manual/metaprogramming/#Macro-invocation-1

phipsgabler replied:
Right... I never used those constants myself, so I just thought that `@__LINE__` does the right thing there -- which id doesn't.  But `__source__.line` gives the location the the call of the `@model` macro, IIUC, which is also the wrong thing.  Ideally, we would extract the original line number line with the tilde. 

I have no idea how to do that in a simple way.  You'd somehow have to `zip(block.args[0:2:end], block.args[1:2:end])` for every block, and pass the `LineNumberNode` to the transformation of the `Expr` after it.  But probably even that is too naive.

phipsgabler replied:
Maybe something like this:

```
julia> function linenumbers(expr::Expr, numbers=IdDict{Any, LineNumberNode}())
           if Meta.isexpr(expr, :block)
               i = 1
               while i <= length(expr.args)
                   if expr.args[i] isa LineNumberNode
                       numbers[expr.args[i+1]] = expr.args[i]
                       linenumbers(expr.args[i], numbers)
                       i += 2
                   else
                       linenumbers(expr.args[i], numbers)
                       i += 1
                   end
               end
           else
               foreach(e -> linenumbers(e, numbers), expr.args)
           end
           return numbers
       end
linenumbers (generic function with 4 methods)

julia> linenumbers(expr, numbers=IdDict{Any, LineNumberNode}()) = numbers
linenumbers (generic function with 4 methods)

julia> linenumbers(:(1 + begin
           x = 1
           2
           end))
IdDict{Any,LineNumberNode} with 2 entries:
  :(x = 1) => :(#= REPL[28]:2 =#)
  2        => :(#= REPL[28]:3 =#)
```

devmotion replied:
I think that's maybe too much for just getting the line numbers correct. I guess a simpler approach would be to not trying to figure out the line number but let Julia handle it. If we do not introduce any new line numbers when generating the tilde expressions, it will correctly point to the line numbers in the user code. As an example:
```julia
julia> x = rand(5);

julia> @model demo(x, n = 5, ::Type{T} = Float64) where {T} = begin
           m = Vector{T}(undef, n)
           @. m ~ rand()
           @. x ~ Normal.(m, 1)
       end

julia> demo(x)()
ERROR: ArgumentError: Right-hand side of a ~ must be subtype of Distribution or a vector of Distributions.
Stacktrace:
 [1] macro expansion at ./REPL[61]:3 [inlined]
 [2] ##evaluator#1411(::DynamicPPL.Model{var"###evaluator#1411",(:x, :n, :T),Tuple{Array{Float64,1},Int64,Type{Float64}},(),DynamicPPL.ModelGen{var"###generator#1412",(:x, :n, :T),(:n, :T),Tuple{Int64,Type{Float64}}}}, ::DynamicPPL.VarInfo{DynamicPPL.Metadata{Dict{DynamicPPL.VarName,Int64},Array{Distribution,1},Array{DynamicPPL.VarName,1},Array{Real,1},Array{Set{DynamicPPL.Selector},1}},Real}, ::DynamicPPL.SampleFromPrior, ::DynamicPPL.DefaultContext) at /home/david/.julia/dev/DynamicPPL/src/compiler.jl:469
 [3] (::DynamicPPL.Model{var"###evaluator#1411",(:x, :n, :T),Tuple{Array{Float64,1},Int64,Type{Float64}},(),DynamicPPL.ModelGen{var"###generator#1412",(:x, :n, :T),(:n, :T),Tuple{Int64,Type{Float64}}}})(::DynamicPPL.VarInfo{DynamicPPL.Metadata{Dict{DynamicPPL.VarName,Int64},Array{Distribution,1},Array{DynamicPPL.VarName,1},Array{Real,1},Array{Set{DynamicPPL.Selector},1}},Real}, ::DynamicPPL.SampleFromPrior, ::DynamicPPL.DefaultContext) at /home/david/.julia/dev/DynamicPPL/src/model.jl:118 (repeats 2 times)
 [4] top-level scope at REPL[62]:1
```
In these experiments the output of the macro is
```julia
julia> @macroexpand @model demo(x, n = 5, ::Type{T} = Float64) where {T} = begin
           m = Vector{T}(undef, n)              
           @. m ~ rand()                                                                                    
           @. x ~ Normal.(m, 1)                                                                                                                                                                                          
       end
quote                          
    #= /home/david/.julia/dev/DynamicPPL/src/compiler.jl:461 =#
    function var"##evaluator#1427"(var"##model#1416"::DynamicPPL.Model, var"##vi#1414"::DynamicPPL.VarInfo, var"##sampler#1415"::AbstractMCMC.AbstractSampler, var"##ctx#1413"::DynamicPPL.AbstractContext)
        #= /home/david/.julia/dev/DynamicPPL/src/compiler.jl:467 =#
        begin                                                                                                                                                                                                            
            x = (DynamicPPL.matchingvalue)(var"##sampler#1415", var"##vi#1414", (var"##model#1416").args.x)
            n = (DynamicPPL.matchingvalue)(var"##sampler#1415", var"##vi#1414", (var"##model#1416").args.n)
            T = (DynamicPPL.matchingvalue)(var"##sampler#1415", var"##vi#1414", (var"##model#1416").args.T)
        end                                                                                                                                                                                                              
        #= /home/david/.julia/dev/DynamicPPL/src/compiler.jl:468 =#
        (DynamicPPL.resetlogp!)(var"##vi#1414")
        #= /home/david/.julia/dev/DynamicPPL/src/compiler.jl:469 =#
        begin
            #= REPL[63]:1 =#                                                                                
            #= REPL[63]:2 =#                                                                                
            m = Vector{T}(undef, n)                                                                         
            #= REPL[63]:3 =#                                                                                                                                                                                                         begin                                                                                                                                                                                                        
                var"##tmpright#1417" = rand.()
                (DynamicPPL.isadist)(var"##tmpright#1417") || throw(ArgumentError("Right-hand side of a ~ must be subtype of Distribution or a vector of Distributions."))
                var"##vn#1419" = (DynamicPPL.VarName){:m}("")          
                var"##inds#1420" = ()                                                                       
                var"##out#1418" = (DynamicPPL.dot_tilde_assume)(var"##ctx#1413", var"##sampler#1415", var"##tmpright#1417", m, var"##vn#1419", var"##inds#1420", var"##vi#1414")
                m .= var"##out#1418"[1]
                (DynamicPPL.acclogp!)(var"##vi#1414", var"##out#1418"[2])
            end                                                                                                                                                                                                          
            #= REPL[63]:4 =#
            begin
                var"##tmpright#1421" = Normal.(m, 1)
                (DynamicPPL.isadist)(var"##tmpright#1421") || throw(ArgumentError("Right-hand side of a ~ must be subtype of Distribution or a vector of Distributions."))
                var"##vn#1423" = (DynamicPPL.VarName){:x}("")
                var"##inds#1424" = ()
                var"##isassumption#1425" = begin
                        let var"##vn#1426" = (DynamicPPL.VarName){:x}("")
                            if !((DynamicPPL.inargnames)(var"##vn#1426", var"##model#1416")) || (DynamicPPL.inmissings)(var"##vn#1426", var"##model#1416")
                                true
                            else
                                if (DynamicPPL.inargnames)(var"##vn#1426", var"##model#1416")
                                    x === missing
                                else
                                    throw("This point should not be reached. Please report this error.")
                                end
                            end
                        end
                    end
                if var"##isassumption#1425"
                    var"##out#1422" = (DynamicPPL.dot_tilde_assume)(var"##ctx#1413", var"##sampler#1415", var"##tmpright#1421", x, var"##vn#1423", var"##inds#1424", var"##vi#1414")
                    x .= var"##out#1422"[1]
                    (DynamicPPL.acclogp!)(var"##vi#1414", var"##out#1422"[2])
                else
                    (DynamicPPL.acclogp!)(var"##vi#1414", (DynamicPPL.dot_tilde_observe)(var"##ctx#1413", var"##sampler#1415", var"##tmpright#1421", x, var"##vn#1423", var"##inds#1424", var"##vi#1414"))
                end
            end
        end
    end
    #= /home/david/.julia/dev/DynamicPPL/src/compiler.jl:472 =#
    var"##generator#1428"(x, n = 5, T::Type{<:Any} = Float64) = begin
            #= /home/david/.julia/dev/DynamicPPL/src/compiler.jl:472 =#
            (DynamicPPL.Model)(var"##evaluator#1427", (DynamicPPL.namedtuple)(NamedTuple{(:x, :n, :T), Tuple{Core.Typeof(x), Core.Typeof(n), Core.Typeof(T)}}, (x, n, T)), (DynamicPPL.ModelGen){(:x, :n, :T)}(var"##generator#1428", (DynamicPPL.namedtuple)(NamedTuple{(:n, :T), Tuple{Core.Typeof(5), Core.Typeof(Float64)}}, (5, Float64))))
        end
    #= /home/david/.julia/dev/DynamicPPL/src/compiler.jl:473 =#
    var"##generator#1428"(; x, n = 5, T::Type{<:Any} = Float64) = begin
            #= /home/david/.julia/dev/DynamicPPL/src/compiler.jl:457 =#
            var"##generator#1428"(x, n, T)
        end
    #= /home/david/.julia/dev/DynamicPPL/src/compiler.jl:475 =#
    demo = (DynamicPPL.ModelGen){(:x, :n, :T)}(var"##generator#1428", (DynamicPPL.namedtuple)(NamedTuple{(:n, :T), Tuple{Core.Typeof(5), Core.Typeof(Float64)}}, (5, Float64)))
end
```

phipsgabler replied:
Oh yeah, that's better.

mohamed82008 commented on src/compiler.jl:
```diff
@@ -11,59 +11,46 @@ function _error_msg()
+                true
             else
-                throw("This point should not be reached. Please report this error.")
+                if $(DynamicPPL.inargnames)($vn, $model)
```
This condition is unnecessary and is implied by the fact that we are in the `else` branch.

devmotion replied:
I didn't check if the branches make sense, I just copied the current implementation :smile:

mohamed82008 replied:
Ya it was a bad on my side.

mohamed82008 commented on src/compiler.jl:
```diff
@@ -11,59 +11,46 @@ function _error_msg()
 """
-macro isassumption(model, expr::Union{Symbol, Expr})
-    # Note: never put a return in this... don't forget it's a macro!
+function isassumption(model, expr::Union{Symbol, Expr})
```
We should probably a check for `Inf` to fix https://github.com/TuringLang/DynamicPPL.jl/issues/15.

devmotion replied:
If we want to add a special case for explicit `Inf` in the model definition to solve https://github.com/TuringLang/DynamicPPL.jl/issues/15, we should check for it while constructing the macro output and not at runtime. Of course, we could still check for `Inf` at runtime as well, if the LHS only evaluates to `Inf`.

I don't really see yet why you might want to write your model in the way  presented in the issue though. Wouldn't it be more natural to write
```julia
res = solve(...)
if !res.success
	@logpdf() = -Inf
	return
end

data ~ dist
```
?

mohamed82008 replied:
True and technically `Inf` can be a variable identifier in a function in Julia so we can close that issue, @xukai92?

mohamed82008 commented on src/compiler.jl:
```diff
@@ -217,7 +204,7 @@ function replace_logpdf!(model_info)
     ex = MacroTools.postwalk(ex) do x
         if @capture(x, @logpdf())
-            :($vi.logp[])
+            :($(DynamicPPL.getlogp)($vi))
```
This is not good because we need to be able to do `@logpdf() += ...`.

devmotion replied:
Ah yes, I forgot about this use case.

I'm still wondering, however, if we actually need this macro. If one provides `@vi`, `@sampler`, `@model`, and `@context` macros that are replaced by the input arguments `vi`, `sampler`, `model`, and `context` of the evaluator function, users have access to pretty much everything in their model definition.

mohamed82008 replied:
True but speaking for the lazy users out there of which I am a proud one, I like the `@logpdf() += ...` syntax :)

devmotion replied:
BTW maybe a bit nicer would be to make `vi`, `sampler`, `model`, and `context` reserved names in a `@model` context and just use these names for the input arguments instead of creating some with `gensym`. Then users could just write `getlogp(vi)` or `vi.logp[]` etc.

mohamed82008 replied:
Well the use case of these variables is very niche so I don't think so many reserved names are justified. Macros are fine I think because people don't define macros as often as they define variables.

devmotion replied:
Sure, macros are less likely collide with any user definitions, but I assume the risk of a user actually defining variables with these names shouldn't be too high (one could even think about using the more explicit and maybe less frequently used `varinfo` instead of `vi`). Alternatively, one could also use, e.g., `_vi`, `_sampler`, `_model`, and `_context` to signal that these are internal variables.

Personally, I'm not a big fan of the macros since their use feels slightly unnatural and I don't like that I have to add brackets all the time. Seems I'm also a lazy person :smile:

mohamed82008 replied:
I agree that macros can be a bit inconvenient but on the plus side, we can guarantee that no user out there will get cryptic errors because he/she overwrote a reserved variable like `varinfo` or otherwise.

devmotion replied:
One could display a warning or information if a user uses these internal variables in his/her models to ensure that nothing magically is happening without them noticing.

devmotion replied:
Though I guess we don't have to come up with a perfect solution for it in this PR, I just revert the logpdf change for now :smiley:

mohamed82008 replied:
👍

mohamed82008 commented on src/compiler.jl:
```diff
@@ -294,45 +281,55 @@ function generate_tilde(left, right, model_info)
+        if vsym(left) in model_info[:args]
+            @gensym isassumption
+            return quote
+                $(top...)
```
Why not just plug in the content of `top` here? It makes it more readable instead of having to jump up to see what `top` is.

devmotion replied:
To avoid code duplication and possible discrepancies between the different cases. There are three different branches that all need `top` (two of them a slightly extended version) and two branches that need `assumption`.

mohamed82008 replied:
Ok I see, makes sense.

mohamed82008 commented on src/compiler.jl:
```diff
@@ -294,45 +281,55 @@ function generate_tilde(left, right, model_info)
+                $(top...)
+                $isassumption = $(DynamicPPL.isassumption(model, left))
+                if $isassumption
+                    $(assumption...)
```
Same comment as `top` above.

mohamed82008 commented on src/compiler.jl:
```diff
@@ -425,42 +427,29 @@ function build_output(model_info)
-            end
-        end)
+        push!(unwrap_data_expr.args,
+              :($var = $(DynamicPPL.matchingvalue)($sampler, $vi, $(model).args.$var)))
```
Cool!

mohamed82008 commented on src/compiler.jl:
```diff
@@ -261,14 +244,14 @@ function replace_tilde!(model_info)
         if dotargs !== nothing
             L, R = dotargs
-            return generate_dot_tilde(L, R, model_info)
+            return Base.remove_linenums!(generate_dot_tilde(L, R, model_info))
```
Why are we removing line numbers? I believe these are used by the stacktrace to show correct line numbers where an error happens.

devmotion replied:
It's to ensure that the stacktraces are actually helpful for users: https://github.com/TuringLang/DynamicPPL.jl/pull/57#discussion_r403928600

mohamed82008 replied:
Nice, thanks for explaining.

mohamed82008 commented:
This PR is beautiful, nice work @devmotion!

phipsgabler commented on src/varname.jl:
```diff
@@ -34,43 +34,34 @@ macro varname(expr::Union{Expr, Symbol})
 
 macro vsym(expr::Union{Expr, Symbol})
-    expr |> vsym
+    return :(QuoteNode($(vsym(expr))))
```
Shouldn't this really be `:($(QuoteNode(vsym(expr))))`, or more simply `QuoteNode(vsym(expr))`?

devmotion replied:
Oh yes, not sure why I did that :smile: Can you just change that in your PR?

phipsgabler replied:
That's exactly what I just did :)
